File: src/main/java/com/bluehabit/eureka/common/BaseResponse.java
/*
 * Copyright © 2023 Blue Habit.
 *
 * Unauthorized copying, publishing of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 */

package com.bluehabit.eureka.common;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;

public final class BaseResponse<T> {
    private final int statusCode;
    private final String message;
    private final T data;

    private BaseResponse(int statusCode, String message, T data) {
        this.statusCode = statusCode;
        this.message = message;
        this.data = data;
    }

    public static <O> ResponseEntity<BaseResponse<O>> failed(int statusCode, String message) {
        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(new BaseResponse<>(statusCode, message, null));
    }

    public static <O> BaseResponse<O> error(int statusCode, String message) {
        return new BaseResponse<>(statusCode, message, null);
    }

    public static <O> ResponseEntity<BaseResponse<O>> success(String message, O data) {
        return ResponseEntity.ok(new BaseResponse<>(HttpStatus.OK.value(), message, data));
    }

    public int getStatusCode() {
        return statusCode;
    }

    public String getMessage() {
        return message;
    }

    public T getData() {
        return data;
    }
}

File: src/main/kotlin/com/bluehabit/budgetku/common/model/BaseResponse.kt
/*
 * Copyright © 2023 Blue Habit.
 *
 * Unauthorized copying, publishing of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 */

package com.bluehabit.budgetku.common.model

import com.bluehabit.budgetku.common.exception.UnAuthorizedException
import com.bluehabit.budgetku.data.permission.Permission
import com.bluehabit.budgetku.data.user.userCredential.UserCredential
import com.bluehabit.budgetku.data.user.userCredential.UserCredentialRepository
import org.springframework.security.core.context.SecurityContextHolder

data class BaseResponse<DATA>(
    var code: Int = 0,
    var data: DATA? = null,
    var message: String = ""
)

data class AuthBaseResponse<DATA>(
    var code: Int = 0,
    var data: DATA? = null,
    var message: String = "",
    var token: String = ""
)

fun <Data> baseResponse(lambda: BaseResponse<Data>.() -> Unit): BaseResponse<Data> =
    BaseResponse<Data>().apply(lambda)

fun <Data> baseAuthResponse(lambda: AuthBaseResponse<Data>.() -> Unit):AuthBaseResponse<Data> =
    AuthBaseResponse<Data>().apply(lambda)

fun <Type> buildResponse(
    userCredentialRepository: UserCredentialRepository,
    allow: (Collection<Permission>) -> Boolean,
    next: (currentUserCredential: UserCredential) -> Type
): Type {
    val email = SecurityContextHolder.getContext().authentication.principal.toString();
    if (email.isEmpty()) throw UnAuthorizedException("[98] You don't have access!")

    val user =
        userCredentialRepository.findByUserEmail(email) ?: throw UnAuthorizedException("[98] You don't have permission")

    if (!allow(user.userPermissions)) {
        throw UnAuthorizedException("[98] You don't have access!")
    }

    return next(user)
}

fun <Type> buildResponse(
    userCredentialRepository: UserCredentialRepository,
    next: (userCredential: UserCredential) -> Type
): Type {
    val email = SecurityContextHolder.getContext().authentication.principal.toString();
    if (email.isEmpty()) throw UnAuthorizedException("[98] You don't have access!")

    val user =
        userCredentialRepository.findByUserEmail(email) ?: throw UnAuthorizedException("[98] You don't have permission")

    return next(user)
}
